### PR TITLE
Generalize PRUCQueue and move it to thread.py

### PR DESCRIPTION
**Pull request by xresende**

### CODE REVIEW

File: siriuspy/siriuspy/thread.py [hunk 1]
```diff
@@ -114,3 +115,95 @@ def stop(self):
...
+
+    def append(self, operation, unique=False):
+        """Append operation to queue."""
+        DequeThread._lock.acquire(blocking=True)
+        if not self._ignore:
+            if not unique:
+                super().append(operation)
+                self._last_operation = operation
+            else:
+                # super().append(operation)
```
Comment (fernandohds564) [inline siriuspy/siriuspy/thread.py]: I think it is not necessary to keep these comment lines.

File: siriuspy/siriuspy/thread.py [hunk 2]
```diff
@@ -114,3 +115,95 @@ def stop(self):
...
+                    self._last_operation = operation
+        DequeThread._lock.release()
+
+    def clear(self):
+        """Clear deque."""
+        self._lock.acquire()
+        super().clear()
+        self._lock.release()
+
+    def popleft(self):
```
Comment (fernandohds564) [inline siriuspy/siriuspy/thread.py]: why not reimplement pop too?

File: siriuspy/siriuspy/thread.py [hunk 3]
```diff
@@ -114,3 +115,95 @@ def stop(self):
...
+        """Pop left operation from queue."""
+        DequeThread._lock.acquire(blocking=True)
```
Comment (fernandohds564) [inline siriuspy/siriuspy/thread.py]: this is ugly, use `with` statement to lock.

File: siriuspy/siriuspy/thread.py [hunk 4]
```diff
@@ -114,3 +115,95 @@ def stop(self):
...
+        if super().__len__() > 0:
+            value = super().popleft()
+        else:
```
Comment (fernandohds564) [inline siriuspy/siriuspy/thread.py]: do you really want to change the behavior of `deque` and not raise a `IndexError` here?

File: siriuspy/siriuspy/thread.py [hunk 5]
```diff
@@ -114,3 +115,95 @@ def stop(self):
...
+        DequeThread._lock.acquire(blocking=True)
+        if not self._ignore:
+            if not unique:
+                super().append(operation)
+                self._last_operation = operation
+            else:
+                # super().append(operation)
+                # self._last_operation = operation
+                n = self.count(operation)
+                if n == 0:
```
Comment (fernandohds564) [inline siriuspy/siriuspy/thread.py]: This is nice!

File: siriuspy/siriuspy/thread.py (hunk 1 above)
Comment (xresende) [reply siriuspy/siriuspy/thread.py]: done.

File: siriuspy/siriuspy/thread.py (hunk 2 above)
Comment (xresende) [reply siriuspy/siriuspy/thread.py]: done

File: siriuspy/siriuspy/thread.py (hunk 3 above)
Comment (xresende) [reply siriuspy/siriuspy/thread.py]: done

File: siriuspy/siriuspy/thread.py (hunk 4 above)
Comment (xresende) [reply siriuspy/siriuspy/thread.py]: done

**Comment by fernandohds564** [comment]
@xresende , I've created a branch dev-packages/PR-update-prucqueue-suggestions with a bug fix and some style suggestions. Please take a look at it before merging this PR.

**Comment by xresende** [comment]
@fernandohds564 , your branch has been tested and merged in this PR. thanks.

**Pull request merged**